### PR TITLE
JIT: fix issue with assert seen in OSR+PGO tests

### LINKED ISSUE #69032: JIT: libraries PGO assert during Find Loops
Many libraries pgo tests now failing with
```
Assert failure(PID 5952 [0x00001740], Thread: 6833 [0x1ab1]): Assertion failed 'gtOper == GT_CNS_INT' in '<>c__DisplayClass35_0:<TestTryUpdate>b__1(System.Object):this' during 'Find loops' (IL size 171; hash 0x83ebfa29; Tier1-OSR)

    File: /__w/1/s/src/coreclr/jit/gentree.h Line: 2123
    Image: /datadisks/disk1/work/9E9D08F3/p/dotnet
```
across various options and architectures.

See
https://dev.azure.com/dnceng/public/_build/results?buildId=1758094&view=ms.vss-test-web.build-test-results-tab

Spot checking all the failures seem to be in Tier1-OSR compiles, so likely this is from https://github.com/dotnet/runtime/pull/68725.

Similar asserts in the jit-experimental tests: https://dev.azure.com/dnceng/public/_build/results?buildId=1758339&view=ms.vss-test-web.build-test-results-tab

### PR DESCRIPTION
**Pull request by AndyAyersMS**
`IsIconHandle` only works if we know the tree is an integer.

Closes #69032.

### CODE REVIEW

**Comment by AndyAyersMS** [comment]
@BruceForstall PTAL
cc @dotnet/jit-contrib

**Review by jakobbotsch — APPROVED** [review]
Is there any reason we couldn't just move these functions into `GenTreeIntCon`?

File: src/coreclr/jit/gentree.cpp [hunk 1]
```diff
@@ -16487,7 +16487,7 @@ bool GenTree::IsBlockProfileUpdate()
 
     GenTree* const addr = lhs->AsIndir()->Addr();
 
-    return addr->IsIconHandle(GTF_ICON_BBC_PTR);
+    return addr->IsIntegralConst() && addr->IsIconHandle(GTF_ICON_BBC_PTR);
```
Comment (SingleAccretion) [inline src/coreclr/jit/gentree.cpp:16490]: I suppose the more technically correct test would be `IsCnsIntOrI` as `IsIntegralConst` will be `true` for `CNS_LNG` as well.

**Comment by AndyAyersMS** [comment]
> Is there any reason we couldn't just move these functions into `GenTreeIntCon`?

You mean the `IsIconHandle` overloads?

**Comment by AndyAyersMS** [comment]
Seems like I should also add a test.

**Comment by jakobbotsch** [comment]
> You mean the `IsIconHandle` overloads?

Yes (for all the functions around here that assert `gtOper == GT_CNS_INT`). Assuming we do not end up with other nodes with the `GT_CNS_INT` oper.

Either that or I think `IsIconHandle` and co. should check for the correct oper on their own. Currently these functions end up being sort of dynamically typed for seemingly no good reason (probably a vestige from before the node inheritance hierarchy existed).

**Comment by AndyAyersMS** [comment]
Ok, let me look into moving those -- depending on what I see for callers I'll either generalize them to work for any tree or move them down off the base class.

**Comment by AndyAyersMS** [comment]
> Ok, let me look into moving those -- depending on what I see for callers I'll either generalize them to work for any tree or move them down off the base class.

There are about 100 uses so seems like generalizing them to work for any tree would be the least disruptive.

**Comment by AndyAyersMS** [comment]
Ok, I made `IsIconHandle` work for all trees. The get/clear methods I left alone.

There is also quite a bit of code that checks for non-handle ints: `if (op1->IsCnsIntOrI() && !op1->IsIconHandle())` which could arguably be encapsulated with yet another predicate helper, but I also left those as is for now.

**Review by BruceForstall — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
Libraries failure looks like #68443.
Timeout on arm64 should be addressed by #69129.

**Pull request merged**